pymc3/distributions/multivariate.py

```diff
@@ -522,7 +522,7 @@ def logp(self, value):
             tt.sum(logpow(value, a - 1) - gammaln(a), axis=-1) + gammaln(tt.sum(a, axis=-1)),
             tt.all(value >= 0),
             tt.all(value <= 1),
-            np.logical_not(a.broadcastable),
+            np.logical_not(a.broadcastable[-1]),
```

Comment:
What is the exact reasoning behind this change?  I can't seem to make sense of it.

Reply:
OK, I think I see what was happening.  The old `Dirichlet.k` was only supposed to be the last dimension of the [`shape` in `Dirichlet.__init__`](https://github.com/pymc-devs/pymc3/blob//pymc3/distributions/multivariate.py#L499).  In other words, the `shape` variable was a misnomer, because it wasn't the actual shape of the `Dirichlet` variable.

Reply:
Thanks for checking this out. I think the above behaviour is expected because -
- Before the commit f07c273, we have a check k > 1, which suggests the shape of last dimension must be at least 2.
- The wiki page https://en.wikipedia.org/wiki/Dirichlet_distribution also supports this idea mentioning that `k>=2`.
- Even the TFP Dirichlet distribution seems to be very strict with `k` being < 2 and raises an error.

https://github.com/tensorflow/probability/blob/43a9d6c3f5992a24ddf7aa3fa826a038d70697c5/tensorflow_probability/python/distributions/dirichlet.py#L349-L350

I just modified the logic to not accept shape of last dimension when it is 1. Since shape (1, ) is broadcastable, so its logical not is False and that's why model.logp evaluates to `-inf`

Reply:
> * Before the commit [f07c273](https://github.com/pymc-devs/pymc3/commit/f07c273954594289c3419bb729ebc80f8e1143e9), we have a check k > 1, which suggests the shape of last dimension must be at least 2.

I don't see anything in the old or new code that prevents `k == 1`.  

Otherwise, you're referring to the logic in question, which gives incorrect results for that degenerate case.  In other words,  `k > 1` isn't really a _check_ in the sense of input validation, it's simply a condition in the log-likelihood logic; unfortunately, it's one that produces the wrong result when `k == 1`.  

Since there isn't a good case for supporting `k == 1`, it looks like we need an actual input validation check to go along with this change.

FYI: `k > 1` is equivalent to `np.logical_not(a.broadcastable)` when `k = a.shape` (or the shape of the variable in general).  The main advantage provided by the latter approach is that `self.shape` is no longer needed in `Dirichlet.logp`, which prevents issues caused by shape changes (e.g. when sampling with out-of-sample data).


> * The wiki page https://en.wikipedia.org/wiki/Dirichlet_distribution also supports this idea mentioning `k>=2`.
> * Also, the TFP Dirichlet distribution seems to be very strict with `k` being < 2 and raises an error.
> 
> https://github.com/tensorflow/probability/blob/43a9d6c3f5992a24ddf7aa3fa826a038d70697c5/tensorflow_probability/python/distributions/dirichlet.py#L349-L350

As a counter example, I believe SciPy and NumPy both allow `k == 1`, and those are what the majority of our code is based on.

Reply:
Thanks for the deep dive on this.